### ISSUE TITLE
Indicate required inputs on Protoflux Nodes

### ISSUE THREAD

**Issue by Banane9** [Yellow-Dog-Man/Resonite-Issues#926]
### Is your feature request related to a problem? Please describe.

Some Protoflux Nodes have inputs that are strictly required for their execution to work, while others can be left as default.
In some cases this can be obvious (e.g. when a node operates on a slot, it clearly needs that input), but in other cases it can lead to confusion when required inputs aren't indicated (see #923). Inputs may also wrongly be considered required by people unfamiliar with protoflux.

### Describe the solution you'd like

Required inputs should be clearly indicated on the nodes, or at the very least on the tooltip shown when using the protoflux tool to point at an input. This could be achieved by giving those inputs a different shape, a different pattern, or something else.

### Describe alternatives you've considered

Leave things potentially confusing for people.

**Comment by Frooxius** [comment]
I'll pick this on when I can. I'll have to see what kind of visual we could go with here, probably tack on an symbol or something on top or next to the name.

Any thoughts/preferences?

**Comment by TisFoolish** [comment]
An asterisk next to required fields and having it say `(required)` in the hover tooltip?

**Comment by Nytra** [comment]
Maybe an exclamation point icon `!`. I think that would be quite intuitive to understand what it means.

**Comment by Nytra** [comment]
> An asterisk next to required fields and having it say `(required)` in the hover tooltip?

A problem with asterisk is that some nodes have this already, for example:

![2023-12-18 18 16 14](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/14206961/1269d2b7-7bc8-40ff-bb8b-2dea2b220b3d)

The call input has a asterisk.

**Comment by Frooxius** [comment]
I wouldn't use asterisk, because that's already used as a meaning for implicit output. And generally the symbol means either multiplication or wildcard. Same with "!" which can indicate negation. We don't want to overload meaning of the symbols, because they can lead to confusion, so ideally we want something distinct. 

Maybe adding brackets or something `<some input>` vs `some input`? But not sure how I feel about that one too.

**Comment by JackTheFoxOtter** [comment]
If the input is required for the node to function, wouldn't it also make sense for this to cause a ProtoFlux compile time error? (Nodes turn red)

**Comment by JackTheFoxOtter** [comment]
Also, for visualization, I'd suggest underlining the variable name, I think that isn't used for any other meaning yet.

**Comment by Frooxius** [comment]
I don't think it should result in compile time error. The node is still operating, it just becomes more of "no-op". The exact behavior will also depend somewhat on the specific node.

For a lot of the nodes the "required" input is more like "you need to provide this if you want the relevant state of inputs to flip/change".

Underline could work maybe, but I'm also worried a bit about it being visually busy and making the text harder to read.

**Comment by JackTheFoxOtter** [comment]
I disagree with the term "required" then.

In other programming languages, if you have required function arguments, not providing those will result in a compile-time (or run-time) exception. If the argument is optional, not providing it might have an impact on the result, but it won't cause an exception.

I'm not sure what other examples there are, but I would argue that if you have a node that's meant to change a state, but it's not given a state to change to, it not compiling / throwing an error makes more sense than it just not doing anything.

**Comment by Banane9** [comment]
> Underline could work maybe, but I'm also worried a bit about it being visually busy and making the text harder to

I could also see making the text *italic* for one kind of parameter work. Probably for required parameters, as they're way rarer - so it keeps its specialness. That would also fit the alternative term "emphasis", which often uses italic - in this case emphasising the required parameters.


> An asterisk next to required fields and having it say `(required)` in the hover tooltip?

Fitting with the italic on the visual's name, this could also just say *Required* in an extra line on the tooltip 🤔 
In the same vein, it would be cool to indicate a defaultable parameter's default value there instead.

**Comment by Frooxius** [comment]
I think we might be over complicating this a fair bit.

Perhaps better word would be that input is "expected". Or that explicit choice is expected, if you want it to do something. 

But the general thing we should focus on is still the same - indication that explicit choice is expected for given input.

**Comment by Banane9** [comment]
> Perhaps better word would be that input is "expected". Or that explicit choice is expected, if you want it to do something.

While perhaps more accurate, I'm not sure if other words are as clear and approachable as "required". Even to someone who is not well versed with programming, "required" clearly indicates that they have to connect something to that input or the node won't work, without having to dive into the ways required and optional arguments work in other languages. Because in the end, Protoflux is its own language as well, and it can behave differently.


(Although a strict mode or something would perhaps be a useful addition, especially if you're working on something where everything must work or it will mess things up)

> If the input is required for the node to function, wouldn't it also make sense for it not being provided to cause a ProtoFlux compile time error? (Nodes turn red)
